### PR TITLE
[base] Support transpositions in LevenshteinDFA.

### PR DESCRIPTION
**Pull request by ygorshenin**
@mpimenov @mgsergio PTAL

### CODE REVIEW

File: base/levenshtein_dfa.cpp [hunk 1]
```diff
@@ -89,49 +99,46 @@ size_t const LevenshteinDFA::kStartingState = 0;
...
 bool LevenshteinDFA::Position::SubsumedBy(Position const & rhs) const
 {
-  if (m_numErrors <= rhs.m_numErrors)
+  if (m_errorsLeft >= rhs.m_errorsLeft)
     return false;
 
-  size_t const u = m_offset < rhs.m_offset ? rhs.m_offset - m_offset : m_offset - rhs.m_offset;
-  size_t const v = m_numErrors - rhs.m_numErrors;
+  if (IsStandard() && rhs.IsStandard())
+    return AbsDiff(m_offset, rhs.m_offset) <= rhs.m_errorsLeft - m_errorsLeft;
```
Comment (ygorshenin) [inline base/levenshtein_dfa.cpp]: No, check above :)

Comment (mgsergio) [inline base/levenshtein_dfa.cpp]: See now.

File: base/levenshtein_dfa.cpp [hunk 2]
```diff
@@ -89,49 +99,46 @@ size_t const LevenshteinDFA::kStartingState = 0;
...
 
 // LevenshteinDFA::Position ------------------------------------------------------------------------
-LevenshteinDFA::Position::Position(size_t offset, uint8_t numErrors)
-  : m_offset(offset), m_numErrors(numErrors)
+LevenshteinDFA::Position::Position(size_t offset, uint8_t errorsLeft, bool transposed)
+  : m_offset(offset), m_errorsLeft(errorsLeft), m_transposed(transposed)
 {
 }
 
 bool LevenshteinDFA::Position::SubsumedBy(Position const & rhs) const
```
Comment (mgsergio) [inline base/levenshtein_dfa.cpp]: Could you pleas add a comment to this function?

Comment (ygorshenin) [inline base/levenshtein_dfa.cpp]: Done

**Comment by ygorshenin** [comment]
PTAL

**Comment by mgsergio** [comment]
LGTM

**Comment by ygorshenin** [comment]
@mpimenov PTAL

File: base/levenshtein_dfa.hpp [hunk 3]
```diff
@@ -10,19 +10,19 @@ namespace strings
...
-// but the size grows exponentially with the number of errors, so be
-// reasonable and don't use this class when the number of errors is
-// too high.
+// insertions and replacements as errors. Transpositions are handled
+// in a quite special way - its assumed that all transformations are
+// applied in parallel, i.e. the Levenshtein distance between "ab" and
+// "bca" is three, not two.  The code is based on the work "Fast
+// String Correction with Levenshtein-Automata" by Klaus U. Schulz and
+// Stoyan Mihov.  For a fixed number of allowed errors and fixed
+// alphabet the construction time and size of automata is O(length of
```
Comment (mpimenov) [inline base/levenshtein_dfa.hpp]: Please don't break lines inside O().

File: base/levenshtein_dfa.cpp [hunk 4]
```diff
@@ -256,7 +276,7 @@ bool LevenshteinDFA::IsValid(State const & s) const
 
 bool LevenshteinDFA::IsAccepting(Position const & p) const
 {
-  return m_size - p.m_offset <= m_maxErrors - p.m_numErrors;
+  return m_size - p.m_offset <= p.m_errorsLeft;
```
Comment (mpimenov) [inline base/levenshtein_dfa.cpp]: Don't you need to check that ```p.IsStandard()``` here?

Comment (ygorshenin) [inline base/levenshtein_dfa.cpp]: Yes, I need. Nice catch, thanks!

File: base/levenshtein_dfa.hpp (hunk 3 above)
Comment (ygorshenin) [inline base/levenshtein_dfa.hpp]: Sorry, but this is the Unix fmt tool.

**Comment by ygorshenin** [comment]
PTAL

**Comment by ygorshenin** [comment]
PTAL

**Comment by mpimenov** [comment]
LGTM

**Pull request merged**